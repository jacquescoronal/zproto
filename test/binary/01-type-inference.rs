use zproto::binary::{DeviceMessage, Port, command::*};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut port = Port::open_serial("/dev/nonexistant")?;
    // Confirm the expected return types by explicitly marking them an INCORRECT type
    let _: DeviceMessage<types::Reset> = port.tx_recv((0, HOME))?;
    let _: DeviceMessage<types::Home> = port.tx_recv((0, untyped::HOME, 0i32))?;
    let _: DeviceMessage<types::ReturnSetting> = port.tx_recv((0, RETURN_SETTING, RETURN_CURRENT_POSITION))?;
    let _: DeviceMessage<types::ReturnSetting> = port.tx_recv((0, RETURN_SETTING, SET_TARGET_SPEED))?;
    Ok(())
}