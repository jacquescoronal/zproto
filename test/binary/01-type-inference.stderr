error[E0308]: `?` operator has incompatible types
 --> test/binary/01-type-inference.rs:6:42
  |
6 |     let _: DeviceMessage<types::Reset> = port.tx_recv((0, HOME))?;
  |                                          ^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `Reset`, found struct `Home`
  |
  = note: `?` operator cannot convert from `DeviceMessage<Home>` to `DeviceMessage<Reset>`
  = note: expected struct `DeviceMessage<Reset>`
             found struct `DeviceMessage<Home>`

error[E0308]: `?` operator has incompatible types
 --> test/binary/01-type-inference.rs:7:41
  |
7 |     let _: DeviceMessage<types::Home> = port.tx_recv((0, untyped::HOME, 0i32))?;
  |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `Home`, found `u8`
  |
  = note: `?` operator cannot convert from `DeviceMessage<u8>` to `DeviceMessage<Home>`
  = note: expected struct `DeviceMessage<Home>`
             found struct `DeviceMessage<u8>`

error[E0308]: `?` operator has incompatible types
 --> test/binary/01-type-inference.rs:8:50
  |
8 |     let _: DeviceMessage<types::ReturnSetting> = port.tx_recv((0, RETURN_SETTING, RETURN_CURRENT_POSITION))?;
  |                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `ReturnSetting`, found struct `ReturnCurrentPosition`
  |
  = note: `?` operator cannot convert from `DeviceMessage<ReturnCurrentPosition>` to `DeviceMessage<ReturnSetting>`
  = note: expected struct `DeviceMessage<ReturnSetting>`
             found struct `DeviceMessage<ReturnCurrentPosition>`

error[E0308]: `?` operator has incompatible types
 --> test/binary/01-type-inference.rs:9:50
  |
9 |     let _: DeviceMessage<types::ReturnSetting> = port.tx_recv((0, RETURN_SETTING, SET_TARGET_SPEED))?;
  |                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `ReturnSetting`, found struct `SetTargetSpeed`
  |
  = note: `?` operator cannot convert from `DeviceMessage<SetTargetSpeed>` to `DeviceMessage<ReturnSetting>`
  = note: expected struct `DeviceMessage<ReturnSetting>`
             found struct `DeviceMessage<SetTargetSpeed>`
